Integrations redesign

TomaszKolek opened the pull request.

smarx commented:
*Automated message from Dropbox CLA bot*

@TomaszKolek, it looks like you've already signed the Dropbox CLA. Thanks!

TomaszKolek commented:
@timabbott First commit should be completely ready, not WIP.

timabbott commented:
I merged the first bit as a79acf854fa2b09d7cdf81b0c7e1555fcfc14ff6, after fixing the fact that the two Google links were broken.  Please test more carefully when doing this kind of refactor!  It's very easy to introduce regressions.

TomaszKolek commented:
@timabbott Next commit ready for merging. What I did there:
1. Created package for every webhook
2. Moved code from `<webhook_name>.py` to `<webhook_name>/view.py`
3. Moved code from `zerver/tests/webhooks/test_<webhook_name>.py` to `<webhook_name>/tests.py`
4. Modified `tools/test-backend.py` by adding `zerver.views.webhooks` to suites list.

TomaszKolek commented on zerver/lib/test_helpers.py:
```diff
@@ -357,7 +357,10 @@ def get_all_templates():
     def is_valid_template(p, n):
         # type: (text_type, text_type) -> bool
-        return not n.startswith('.') and not n.startswith('__init__') and isfile(p)
+        return 'webhooks' not in p\
```
please take a look especially at it

TomaszKolek commented:
@timabbott Its probably ready (just wait for tests).

Github tracked that tests are renamed instead of deleting and creating so tracking these changes should be relatively easy. 
It isn't in docs ;-( I clicked every integration on master and integrations-redesign branch and compared - should be ok.

Please especially look at changes like `tools/test-backend`

TomaszKolek commented:
I see 8 tasks left to finish integrations part of GCI - maybe then we'd update the PR and merge it?

timabbott commented:
We'll probably be adding more integrations; it's not really a big deal (despite what GitHub, thinks, it rebases cleanly).  I'm mostly just procrastinating merging this because I haven't decided which directory we want to send this stuff to -- `webhooks/`, `zerver/webhooks/`, or something else.

TomaszKolek commented:
@timabbott Its reviewable again.
It'd be cool to review/merge it as first thing - I was only resolving conflicts during last few hours ;-)

TomaszKolek commented on docs/integration-guide.md:
```diff
@@ -32,10 +32,9 @@ products, ordered here by which types we prefer to write:
-   https://github.com/zulip/zulip/pull/324.
+   For these, you usually just need to create a new directory in the
+   `zerver/webhooks/` directory.  An example commit implementing
+   a new webhook is: https://github.com/zulip/zulip/pull/324.
```
Actually we don't have such PR right now ;-(

TomaszKolek commented:
@timabbott:
I revert changes in integrations.html ( I mean formatting changes)
I added 3 or 4 exceptions to check_template tests because of `pre` tags
I modified guide - please take a look if its correct.

timabbott commented:
Merged, thanks @TomaszKolek!  I noticed a few typos in the integration guide work where you didn't update the commands and paths correctly.  I fixed all the ones I found, but can you do an audit of actually running all the commands and `ls`ing the paths to make sure they're actually correct?  Best that we find the bug rather than a user who doesn't know the system.
